uniqueserial/src/uniqueserial.cpp

```diff
@@ -1,50 +1,89 @@
-#include <unordered_map>
-#include "sphinxudf.h"
 #include "uniqueserial.h"
 
+/**
+ * Расчет коэфф. в зависимости от входного параметра, в частном случае
+ * параметром может выступать коэфициент пакета поделенный на 10
+ *
+ * Для данного c (company_id например) помечаем всподряд идущие строки p раз одним и тем же коэфициентом
+ * если превысили порог p коэфициент увеличиваем на 1 и снова помечаем всподряд идущие строки выборки
+ * одним и тем же коэфициентом.
+ *
+ * Тем самым "вытягиваем" по p строк для каждого c вверх чередуя их по группам;
+ * получается почти честный листинг (товаров):
+ *
+ * company1  packet3   1
+ * company1  packet3   1
+ * company1  packet3   1
+ * company2  packet3   1
+ * company2  packet3   1
+ * company2  packet3   1
+ * company1  packet3   2
+ * company1  packet3   2
+ * company1  packet3   2
+ * company2  packet3   2
+ * и т.д.
+ *
+ * Функция использует блокировку из-за того, что сфинкс может распараллелить один запрос на несколько потоков:
+ * См. https://github.com/manticoresoftware/manticoresearch/issues/87 и документацию.
+ *
+ **/
+sphinx_uint64_t Factor::calculate_factor(unsigned int p) {
+  std::lock_guard<std::mutex> guard(factor_mutex);
+
+  if (counter >= p) {
+    counter = 1;
+    return factor++;
+  } else {
+    counter++;
+    return factor;
+  }
+}
+
```

Comment:
Каждый раз что-ли мютекс инициализируется ?
Наверно накладно это, Томат вроде по-другому предлагал:
> You might init lock along with data at udf_init function provided and do not forget to clean up data at udf_deinit function to prevent memory and resource leaks.

Захватывать мютекс на каждый инкремент наверно тоже дорого, какую-то бы более легкую блокировку использовать, в теории. Ну понятно что щас это вредно ибо преждевременная оптимизация. Но померить влияние этой функции на перфоманс листинга надо померить.

Reply:
на тестовых просадов я не заметил. В проде собираемя на одном регионе тестировать сначала в Барнауле.